aiven/client/argx.py

```diff
@@ -196,47 +260,100 @@ def print_response(
-                file=file,
-            )
-        elif csv:
-            fields = []
-            for field in table_layout:
-                if isinstance(field, str):
-                    fields.append(field)
-                else:
-                    fields.extend(field)
-
-            writer = csvlib.DictWriter(file, extrasaction="ignore", fieldnames=fields)
-            if header:
-                writer.writeheader()
-            for item in result:
-                writer.writerow(item)
-        else:
-            if single_item:
-                result = [result]
+            prepared_data = pretty.prepare_table(result, table_layout)
+            # This code supports the format arg exactly as currently used.
+            if not isinstance(result, (Mapping, Sequence)) or isinstance(result, (basestring, )):
+                raise NotImplementedError(
+                    "Cannot output with format non-mapping or sequence types types: got {result_type}".format(
+                        result_type=type(result)
+                    )
+                )
+            for table_row in prepared_data.table_rows:
+                output_stream.write(format.format(**table_row))
+                output_stream.write("\n")
 
+        if output_format in (OutputFormats.TABLE, OutputFormats.TABLE_NOHEADER):
             pretty.print_table(
                 result,
                 drop_fields=drop_fields,
                 table_layout=table_layout,
-                header=header,
-                file=file,
+                header=output_format == OutputFormats.TABLE,
+                file=output_stream
+            )
+
+        elif output_format in (OutputFormats.CSV, OutputFormats.CSV_NOHEADER):
+            prepared_data = pretty.prepare_table(result, table_layout)
+            fields = []
+            if len(prepared_data.vertical_fields) != 0:
+                fields.append("")
+
+            writer = csvlib.DictWriter(output_stream, extrasaction="ignore", fieldnames=fields)
+
+            if output_format == OutputFormats.CSV:
+                writer.writeheader()
+
+            for row in prepared_data.table_rows:
+                row_item = {k: pretty.format_item(v) for k, v in row.items()}
+                row_item[""] = "\n".join(pretty.yield_vertical_fields(row_item, prepared_data.vertical_fields))
+                writer.writerow(row_item)
+
+        elif output_format == OutputFormats.JSON:
+            jsonlib.dump(result, output_stream, indent=4, sort_keys=True, cls=pretty.CustomJsonEncoder)
+            output_stream.write("\n")
+        elif output_format == OutputFormats.JSON_COMPACT:
+            jsonlib.dump(result, output_stream, sort_keys=True, cls=pretty.CustomJsonEncoder)
+            output_stream.write("\n")
+        elif output_format == OutputFormats.YAML:
+            yamllib.round_trip_dump(result, output_stream, default_flow_style=False)
+
+        else:
+            raise NotImplementedError(
```

Comment:
ValueError would be more appropriate IMO - https://docs.python.org/3/library/exceptions.html

Reply:
Disagree on this one: the entry point to this function via the enum means this code should never be reached. If you do reach it, it's because you extended the enum without adding the handler.

Reply:
I'd say that the enum passed in is a value that cannot be handled, therefore ValueError. I guess you could argue either point, however I don't think this case matches the intent for NotImplementedError:

>  In user defined base classes, abstract methods should raise this exception when they require derived classes to override the method, or while the class is being developed to indicate that the real implementation still needs to be added.

Reply:
Fixed.